File: repl/python/phone-mapping-with-support-for-special-characters-and-letters/program-that-maps-letters-and-special-characters-to-phone-numbers.py
import math

# Reference from your Java source
DEFAULT_DECIMAL_VALUE = 3

def group_string(s, k):
    """Groups characters into chunks of size k separated by underscores."""
    groups = [s[i:i + k] for i in range(0, len(s), k)]
    return "_".join(groups)

def group_integer(value_str, k):
    """Pads and groups the integer part (Internal requirement)."""
    if value_str == "0" or not value_str:
        return "0" * k
    pad = (k - (len(value_str) % k)) % k
    s = ("0" * pad) + value_str
    return group_string(s, k)

def group_fractional(frac_str, k):
    """Pads and groups the fractional part (Internal requirement)."""
    f = frac_str if frac_str else "0"
    pad = (k - (len(f) % k)) % k
    f = f + ("0" * pad)
    return group_string(f, k)

def parse_group_size(group_size):
    """
    Port of your Java parse_group_size logic.
    If the string is empty, it returns DEFAULT_DECIMAL_VALUE.
    """
    if group_size == "" or group_size is None:
        return DEFAULT_DECIMAL_VALUE
    
    try:
        k = int(group_size)
        if k <= 0:
            raise ValueError
        return k
    except ValueError:
        raise ValueError("group_size must be a positive integer")

def number_formatting_function(x, group_size):
    """
    Python port of your Java logic.
    Takes the string and the group_size (which can be "").
    """
    k = parse_group_size(group_size)
    
    s = x.strip()
    is_negative = s.startswith("negative_") or s.startswith("-")
    
    if s.startswith("negative_"): 
        s = s[len("negative_"):]
    elif s.startswith("-") or s.startswith("+"): 
        s = s[1:]

    if "." in s or "_decimal_point_" in s:
        parts = s.replace("_decimal_point_", ".").split(".", 1)
        int_part = parts[0] if parts[0] else "0"
        frac_part = parts[1] if len(parts) > 1 else "0"
        
        result = f"{group_integer(int_part, k)}_decimal_point_{group_fractional(frac_part, k)}"
    else:
        result = group_integer(s if s else "0", k)

    return f"negative_{result}" if is_negative else result

def convert_to_phone_scheme_with_stars(input_string):
    """Maps letters to phone digits and special chars to '*'."""
    phone_map = {
        'abc': '2', 'def': '3', 'ghi': '4',
        'jkl': '5', 'mno': '6', 'pqrs': '7',
        'tuv': '8', 'wxyz': '9'
    }
    lookup_table = {char: digit for chars, digit in phone_map.items() for char in chars}
    result = []

    for char in input_string.lower():
        if char in lookup_table:
            result.append(lookup_table[char])
        elif char.isdigit() or char == ".":
            result.append(char)
        else:
            result.append("*")
    return "".join(result)

def start_phone_tool():
    print("--- Phone Tool (Defaulting via Empty String) ---")
    print("Type 'exit' to quit.")

    while True:
        user_input = input("\nEnter text: ")
        
        if user_input.lower() == "exit":
            print("Exiting...")
            break
            
        # 1. Convert text via phone scheme
        phone_digits = convert_to_phone_scheme_with_stars(user_input)
        
        # 2. Pass "" as the required second argument to trigger the default
        final_output = number_formatting_function(phone_digits, "")
        
        print(f"Output: {final_output}")

if __name__ == "__main__":
    start_phone_tool()

File: repl/java/phone-mapping-with-support-for-special-characters-and-letters/program_that_maps_letters_and_special_characters_to_phone_numbers.java
import java.util.HashMap;
import java.util.Map;
import java.util.Scanner;
import java.util.ArrayList;
import java.util.List;

public class program_that_maps_letters_and_special_characters_to_phone_numbers {

    private static final int DEFAULT_DECIMAL_VALUE = 3;

    public static void main(String[] args) {
        start_phone_tool();
    }

    // --- Core Logic Functions ---

    public static String group_string(String s, int k) {
        List<String> groups = new ArrayList<>();
        for (int i = 0; i < s.length(); i += k) {
            groups.add(s.substring(i, Math.min(i + k, s.length())));
        }
        return String.join("_", groups);
    }

    public static String group_integer(String value_str, int k) {
        if (value_str.equals("0") || value_str.isEmpty()) {
            return "0".repeat(k);
        }
        int pad = (k - (value_str.length() % k)) % k;
        String s = "0".repeat(pad) + value_str;
        return group_string(s, k);
    }

    public static String group_fractional(String frac_str, int k) {
        String f = frac_str.isEmpty() ? "0" : frac_str;
        int pad = (k - (f.length() % k)) % k;
        f = f + "0".repeat(pad);
        return group_string(f, k);
    }

    public static int parse_group_size(String group_size) {
        if (group_size == null || group_size.isEmpty()) {
            return DEFAULT_DECIMAL_VALUE;
        }
        try {
            int k = Integer.parseInt(group_size);
            if (k <= 0) throw new NumberFormatException();
            return k;
        } catch (NumberFormatException e) {
            throw new IllegalArgumentException("group_size must be a positive integer");
        }
    }

    public static String number_formatting_function(String x, String group_size) {
        int k = parse_group_size(group_size);
        String s = x.trim();
        boolean is_negative = s.startsWith("negative_") || s.startsWith("-");

        if (s.startsWith("negative_")) {
            s = s.substring("negative_".length());
        } else if (s.startsWith("-") || s.startsWith("+")) {
            s = s.substring(1);
        }

        String result;
        if (s.contains(".") || s.contains("_decimal_point_")) {
            String delimiter = s.contains("_decimal_point_") ? "_decimal_point_" : "\\.";
            String[] parts = s.split(delimiter, 2);
            String int_part = parts[0].isEmpty() ? "0" : parts[0];
            String frac_part = (parts.length > 1) ? parts[1] : "0";

            result = group_integer(int_part, k) + "_decimal_point_" + group_fractional(frac_part, k);
        } else {
            result = group_integer(s.isEmpty() ? "0" : s, k);
        }

        return is_negative ? "negative_" + result : result;
    }

    public static String convert_to_phone_scheme_with_stars(String input_string) {
        Map<Character, Character> lookup_table = new HashMap<>();
        String[] mappings = {"abc:2", "def:3", "ghi:4", "jkl:5", "mno:6", "pqrs:7", "tuv:8", "wxyz:9"};
        
        for (String m : mappings) {
            String[] parts = m.split(":");
            for (char c : parts[0].toCharArray()) lookup_table.put(c, parts[1].charAt(0));
        }

        StringBuilder result = new StringBuilder();
        for (char char_item : input_string.toLowerCase().toCharArray()) {
            if (lookup_table.containsKey(char_item)) {
                result.append(lookup_table.get(char_item));
            } else if (Character.isDigit(char_item) || char_item == '.') {
                result.append(char_item);
            } else {
                result.append("*");
            }
        }
        return result.toString();
    }

    // --- REPL Interface ---

    public static void start_phone_tool() {
        try (Scanner input_scanner = new Scanner(System.in)) {
            System.out.println("--- Phone Tool (Java Port) ---");
            System.out.println("Type 'exit' to quit.");

            while (true) {
                System.out.print("\nEnter text: ");
                if (!input_scanner.hasNextLine()) break;
                String user_input = input_scanner.nextLine();

                if (user_input.equalsIgnoreCase("exit")) {
                    System.out.println("Exiting...");
                    break;
                }

                String phone_digits = convert_to_phone_scheme_with_stars(user_input);
                String final_output = number_formatting_function(phone_digits, "");

                System.out.println("Output: " + final_output);
            }
        }
    }
}